### PR TITLE
Listing updates august

### PR DESCRIPTION
**Pull request by treygrainger**
- Improves formatting when rendering judgments
- Makes listing code formatting consistent with manuscript formatting across entire codebase.
- Standardizes print statements and now explicitly prints the last item in listings. (tons of new print functions)
- Updates many variables and function names to better suite manuscript
- Cleans up collection rerank interface
- Ch3: Alters term frequency calculation and (tfidf)
- Ch6: Fixes bug in listing 6.18
- ch5: Co ref implementation stubbed in favor of quantiation until isolated dependency environment can be set up
- Ch9: 95/5 test train split + correctly train on training_data
- Ch9: Split notebook
- skg.generate_request => transform_request

Listings 5.1 and 7.14 currently stubbed due to un-resolved environmental  constraints

### CODE REVIEW

File: chapters/ch14/3.question-answering-fine-tuning.ipynb [hunk 1]
```diff
@@ -814,359 +711,21 @@
...
...u001b[0m     args\u001b[38;5;241m=\u001b[39mtraining_args,                                   \u001b[38;5;66;03m# training arguments, defined above\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     36\u001b[0m     eval_dataset\u001b[38;5;241m=\u001b[39mtokenized_datasets[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtest\u001b[39m\u001b[38;5;124m\"\u001b[39m]               \u001b[38;5;66;03m# evaluation dataset\u001b[39;00m\n\u001b[1;32m     37\u001b[0m )\n\u001b[0;32m---> 39\u001b[0m training_outputs \u001b[38;5;241m=\u001b[39m \u001b[43mtrainer\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtrain\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     40\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mTraining Loss:\u001b[39m\u001b[38;5;124m\"\u001b[39m, training_outputs\u001b[38;5;241m.\u001b[39mtraining_loss)\n\u001b[1;32m     41\u001b[0m evaluation_outputs \u001b[38;5;241m=\u001b[39m trainer\u001b[38;5;241m.\u001b[39mevaluate(eval_dataset\u001b[38;5;241m=\u001b[39mtokenized_datasets[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvalidation\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n",
```
Comment (treygrainger) [inline chapters/ch14/3.question-answering-fine-tuning.ipynb:1159]: Need to run notebook to generate clean output before we can merge into `main`

File: chapters/ch04/1.setting-up-the-retrotech-dataset.ipynb [hunk 2]
```diff
@@ -350,8 +344,20 @@
...
+     ]
+    },
+    {
+     "ename": "KeyboardInterrupt",
+     "evalue": "",
+     "output_type": "error",
+     "traceback": [
+      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
+      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
+      "Cell \u001b[0;32mIn[8], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m signals_collection \u001b[38;5;241m=\u001b[39m engine\u001b[38;5;241m.\u001b[39mcreate_collection(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msignals\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m----> 2\u001b[0m \u001b[43msignals_collection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfrom_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdata/retrotech/signals.csv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
```
Comment (treygrainger) [inline chapters/ch04/1.setting-up-the-retrotech-dataset.ipynb:357]: run again so outputs are correct

File: chapters/ch07/2.semantic-search.ipynb [hunk 3]
```diff
@@ -863,59 +861,137 @@
...
-       "  {'business_name': 'Hibiscus', 'score': 3.7260172},\n",
-       "  {'business_name': 'Cho Won Garden', 'score': 3.7260172},\n",
-       "  {'business_name': 'Kojan Bistro', 'score': 3.7260172},\n",
-       "  {'business_name': 'Bulgogi Box', 'score': 3.7260172},\n",
-       "  {'business_name': 'Korean Restaurant', 'score': 3.7260172},\n",
-       "  {'business_name': 'China Wing', 'score': 3.7260172},\n",
-       "  {'business_name': 'Fujiyama', 'score': 3.4513955},\n",
-       "  {'business_name': 'PePeRo', 'score': 3.4513955},\n",
-       "  {'business_name': \"Let's Meat Kbbq\", 'score': 3.4513955}]}"
+       "{'docs': [{'id': 'WnLhd38sH80ViWwzyF7yoA',\n",
```
Comment (treygrainger) [inline chapters/ch07/2.semantic-search.ipynb:881]: TODO: check this to verify why all of these fields are now coming back.

File: chapters/ch08/1.signals-boosting.ipynb [hunk 4]
```diff
@@ -338,19 +357,18 @@
...
-      "Wiping \"normalized_signals_boosts\" collection\n",
-      "Creating \"normalized_signals_boosts\" collection\n",
-      "Status: Success\n",
-      "Successfully written 152252 documents\n",
-      "Signals Aggregation Completed!\n"
+     "ename": "NameError",
+     "evalue": "name 'signals_collection' is not defined",
+     "output_type": "error",
+     "traceback": [
+      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
```
Comment (treygrainger) [inline chapters/ch08/1.signals-boosting.ipynb:368]: run again so outputs are correct

File: chapters/ch12/1.ab-testing-to-active-learning.ipynb [hunk 5]
```diff
@@ -3423,977 +3420,26 @@
...
-      "File \u001b[0;32m/opt/conda/lib/python3.10/warnings.py:467\u001b[0m, in \u001b[0;36mcatch_warnings.__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    465\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_showwarning \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_module\u001b[38;5;241m.\u001b[39mshowwarning\n\u001b[1;32m    466\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_showwarnmsg_impl \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_module\u001b[38;5;241m.\u001b[39m_showwarnmsg_impl\n\u001b[0;32m--> 467\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_record:\n\u001b[1;32m    468\u001b[0m     log \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m    469\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_module\u001b[38;5;241m.\u001b[39m_showwarnmsg_impl \u001b[38;5;241m=\u001b[39m log\u001b[38;5;241m.\u001b[39mappend\n",
-      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
+      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
```
Comment (treygrainger) [inline chapters/ch12/1.ab-testing-to-active-learning.ipynb:3432]: run again so outputs are correct

Comment (dcrouch26) [reply chapters/ch12/1.ab-testing-to-active-learning.ipynb:3449]: Resolved

**Pull request merged**